rust/protocol/src/crypto.rs

```diff
@@ -1,156 +1,244 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2022 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
-use std::convert::TryInto;
-use std::result::Result;
+//! Internal convenience wrappers for symmetric cryptographic primitives.
 
 use aes::cipher::{NewCipher, StreamCipher};
 use aes::{Aes256, Aes256Ctr};
+use arrayref::array_ref;
 use block_modes::block_padding::Pkcs7;
 use block_modes::{BlockMode, Cbc};
+use displaydoc::Display;
 use hmac::{Hmac, Mac, NewMac};
 use sha2::Sha256;
 use subtle::ConstantTimeEq;
+use thiserror::Error;
 
-#[derive(Debug)]
-pub(crate) enum EncryptionError {
-    /// The key or IV is the wrong length.
-    BadKeyOrIv,
-}
+/// bad ciphertext: {0}
+///
+/// Either the input is malformed, or the MAC doesn't match on decryption.
+///
+/// These cases should not be distinguished; message corruption can cause either problem.
+#[derive(Debug, Display, Error)]
+#[ignore_extra_doc_attributes]
+pub struct DecryptionError(pub &'static str);
```

Comment:
Nitpick: For some reason displaydoc on structs weirds me out; maybe it's because enum cases often get abbreviated docs anyway in a way that types don't. Mind switching to thiserror-style for the Display representation instead?